Desktop: Go To Anything by body

anjulalk opened the pull request:
Closes #2683 

Using `/` prefix followed by a search query now shows list of notes which matches the query in their bodies. I had to add a new state variable `highlightKeywords` to prevent body search keywords getting highlighted in the titles.

![1](https://user-images.githubusercontent.com/26695184/76160453-6bad0300-6150-11ea-8d7b-d62020d66ba0.gif)

I've also edited the help text. Hope that's not an issue for translation?

laurent22 commented:
Thanks for the pull request @anjulalk. Could you go a step further and display fragments of the note body below the title, with highlighted keywords?

For example, if the title is "Welcome to Joplin" with body "Joplin is a free, open source note taking and to-do application, which helps you write and organise your notes, and synchronise them between your devices.", and you search for "note". It would display this:

> Welcome to Joplin (note title)
> ... open source **note** taking and ... organise your **note**s, and synchronise ... (note body)
> Welcome (notebook)

So title and notebook stay the same, but when searching within the content, you would display a line in between showing the highlighted results. You don't need to display all the results, maybe just the first 2 or 3.

anjulalk commented:
That'd be a great addition indeed. I've came up with a way to implement this, but I don't know if it's the best way. Therefore please give your feedback.

The results from the search engine currently don't include note bodies. So we can conditionally include the note bodies. (Only if user is searching body. Otherwise note bodies are not included.) Then we can use regular expressions to find locations which matches the query, and then display their surrounding content as you've mentioned. 

The problem is it's not very efficient. Is there a better way? Can we get the matching locations from SQL?

laurent22 commented:
> The problem is it's not very efficient. Is there a better way? Can we get the matching locations from SQL?

Indeed I'd be worried there's potentially a lot of content to be returned. Could you for example only load the body from the first 100 notes from GotoAnything.jsx, and display this? Then beyond 50 notes, you simply display "..." in the results to show that it's not a bug, but we just aren't loading this.

You can use `Note.byIds(noteIds, { fields: ["id", "body"]})` to load only the first 100 notes, based on the IDs you got from the search engine.

anjulalk commented:
Great! I'm working on it. I'll give an update soon.

laurent22 commented:
Any update?

anjulalk commented:
Sorry, it took longer than I expected. I've managed to implement the functionality but I think it may not be perfect. Therefore please give feedback.

![Screenshot at 2020-03-15 22-27-05](https://user-images.githubusercontent.com/26695184/76706291-56197980-670c-11ea-82bb-5e1a9b493a94.png)

For each keyword match, it finds beginning and end indices of the word containing the keyword. (Actually it finds the ending of the word which is 15 characters further than the actual keyword, to get some more words) For example, searching for 'not' gives the string `note taking and to-do`. It saves these indices.

There can be multiple indices overlapping, for example, `note taking and to-do` for `note` keyword, and `taking and to-do application` for `taking` keyword. Then it merges these indices into one: `note taking and to-do application`.

Finally it strips these strings from the body, and join them with ` ... `

anjulalk commented:
Example: fragments for `note` and `taking` keywords are merged
![image](https://user-images.githubusercontent.com/26695184/76706747-fcb34980-670f-11ea-8296-a9918c806afe.png)

laurent22 commented on ElectronClient/plugins/GotoAnything.jsx:
```diff
@@ -165,9 +186,55 @@ class Dialog extends React.PureComponent {
+						}
+
+						fragments = stack.map(f => body.slice(f[0], f[1])).join(' ... ');
+					}
```
Could you refactor this into a function to make it clearer what it does?

anjulalk replied:
I've refactored the merge code into a function and added comments to make it more understandable.

laurent22 commented on ReactNativeClient/lib/string-utils.js:
```diff
@@ -264,6 +264,15 @@ function substrWithEllipsis(s, start, length) {
+	else end += match.index + match[0].length + offset;
+
+	return [begin, end];
+}
```
Please add a comment to describe:
- What this function does
- What are the parameters
- What is the output

anjulalk replied:
I've changed this function to nextWhitespaceIndex(s, begin) which seeks the next whitespace character from `begin` index in string `s` and return its index. It's much more simpler. Let me know if I should make any changes.

anjulalk commented:
Thanks a lot for the review. I've made the requested changes and added more comments. 
The code should be more cleaner and simpler now. Please have a look.

Please let me know of any required changes. I'll make them right away!

laurent22 commented:
Thanks @anjulalk, code looks good now and your comments make it clear how the loop to build up the text fragments works.

Could you just add unit tests for mergeOverlappingIntervals and nextWhitespaceIndex please? I guess you can move mergeOverlappingIntervals to ArrayUtils as it will make it easier to test it. You'll find the relevant test units in tests/ArrayUtils.js and tests/StringUtils.js

anjulalk commented:
I've moved the `mergeOverlappingIntervals()` to `ArrayUtils.js` file, and added unit tests for `mergeOverlappingIntervals` and `nextWhitespaceIndex()` functions. 
Please have a look and give feedback. Thanks!

laurent22 commented on CliClient/tests/ArrayUtils.js:
```diff
@@ -49,4 +49,29 @@ describe('ArrayUtils', function() {
 	}));
 
+	it('should merge overlapping intervals', asyncTest(async () => {
+		const testCases = [
```
Could you add a test case for the empty array please?

laurent22 commented on CliClient/tests/StringUtils.js:
```diff
@@ -41,4 +41,18 @@ describe('StringUtils', function() {
 
+	it('should find the next whitespace character', asyncTest(async () => {
+		const str = 'Joplin is a free, open source\n note taking and *to-do* application';
+		const testCases = [[0, 6], [12, 17], [23, 29], [48, 54]];
```
Likewise, please could you add two test cases:
- for str = ""
- and for when there's no whitespace after the specified index

anjulalk commented:
Apologies, I should've checked those edge cases before. I've added them and fixed the error caused by `mergeOverlappingIntervals()` when providing an empty array. Please have a look.

laurent22 commented:
Thanks for the update @anjulalk, it's good to go now.
